File: src/test/baekjoon/python/queue/RotatingQueue.py
from collections import *
N, M = map(int, input().split())
nums = deque([i for i in range(1, N + 1)])
pick_nums = list(map(int, input().split()))
count = 0
for pick_num in pick_nums:
    if nums[0] == pick_num:
        nums.popleft()
        continue
    index, half_index = nums.index(pick_num), 0

    if len(nums) % 2 == 0:
        half_index = len(nums) // 2 - 1
    else:
        half_index = len(nums) // 2

    if index > half_index:
        while nums[0] != pick_num:
            nums.appendleft(nums.pop())
            count += 1
    else:
        while nums[0] != pick_num:
            nums.append(nums.popleft())
            count += 1
    nums.popleft()
print(count)

File: src/test/baekjoon/java/queue/RotatingQueue.java
package test.baekjoon.java.queue;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.LinkedList;

public class RotatingQueue {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        String[] s = br.readLine().split(" ");
        int N = Integer.parseInt(s[0]);
        int M = Integer.parseInt(s[1]);
        int count = 0;

        LinkedList<Integer> deque = new LinkedList<>();
        for(int i = 1; i < N + 1; i++) {
            deque.add(i);
        }

        String[] nums = br.readLine().split(" ");
        for(int i = 0; i < M; i++) {
            int index = 0;
            int halfIndex = 0;
            int pickNum = Integer.parseInt(nums[i]);
            if(deque.peekFirst() == pickNum) {
                deque.pollFirst();
                continue;
            }

            index = deque.indexOf(pickNum);

            if(deque.size() % 2 == 0) {
                halfIndex = deque.size() / 2 - 1;
            } else {
                halfIndex = deque.size() / 2;
            }

            if(index > halfIndex) {
                while(deque.peekFirst() != pickNum) {
                    count++;
                    deque.addFirst(deque.pollLast());
                }
            } else {
                while(deque.peekFirst() != pickNum) {
                    count++;
                    deque.addLast(deque.pollFirst());
                }
            }
            deque.pollFirst();
        }
        System.out.println(count);
    }
}
